### PR TITLE
Sanitize percentile_approx() output for empty input

### LINKED ISSUE #10856: [FEA] cudf::tdigest::percentile_approx() will return unsanitary lists for empty inputs.
If all of the tdigest inputs to `percentile_approx` are empty, it will return a column containing null rows, as expected, but they will have unsanitary offsets.  

An example where 3 percentiles are requested (3 output values per row expected). 
```
List<double>:
Length : 3
Offsets : 0, 3, 6, 9
Null count: 3
000
   0, 0, 0, 0, 0, 0, 0, 0, 0
```

Ideally this would return

```
List<double>:
Length : 3
Offsets : 0, 0, 0, 0
Null count: 3
000
```

### PR DESCRIPTION
**Pull request by SrikarVanavasam**
Closes #10856 

If all of the tdigest inputs to percentile_approx are empty, it will return a column containing null rows, as expected, but they will have unsanitary offsets. This PR checks if all the inputs are empty and returns an empty column as expected.

### CODE REVIEW

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 1]
```diff
@@ -351,13 +351,14 @@ std::unique_ptr<column> percentile_approx(tdigest_column_view const& input,
   // output is a list column with each row containing percentiles.size() percentile values
   auto offsets = cudf::make_fixed_width_column(
     data_type{type_id::INT32}, input.size() + 1, mask_state::UNALLOCATED, stream, mr);
-  auto row_size_iter = thrust::make_constant_iterator(percentiles.size());
+  bool empty_input   = thrust::reduce(input.size_begin(), input.size_begin() + input.size()) == 0;
```
Comment (ttnghia) [inline cpp/src/quantiles/tdigest/tdigest.cu:354]: This is not safe. The size vales may have something like `-2, 2, -3, 3, ...` which sum to zero.

Let's check if all the size values are zero:
```suggestion
  bool empty_input   = thrust::count_if(rmm::exec_policy(), []__device__() { size[i] == 0} ) == input.size();
```

Comment (ttnghia) [reply cpp/src/quantiles/tdigest/tdigest.cu:354]: If the input is guaranteed to have size initialized (i.e., the size values are all valid) then `thrust::reduce` is enough. But don't forget to add `rmm::exec_policy`.

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 2]
```diff
@@ -351,13 +351,17 @@ std::unique_ptr<column> percentile_approx(tdigest_column_view const& input,
   // output is a list column with each row containing percentiles.size() percentile values
   auto offsets = cudf::make_fixed_width_column(
     data_type{type_id::INT32}, input.size() + 1, mask_state::UNALLOCATED, stream, mr);
-  auto row_size_iter = thrust::make_constant_iterator(percentiles.size());
+  bool empty_input   = thrust::count_if(rmm::exec_policy(stream),
```
Comment (mythrocks) [inline cpp/src/quantiles/tdigest/tdigest.cu:354]:
```suggestion
  auto const empty_input   = thrust::count_if(rmm::exec_policy(stream),
```
Please forgive my quoting [the Good Book](https://www.goodreads.com/book/show/105125.Effective_C_): Item 3 suggests: "Use const wherever possible".
By default, consider making variables `auto const`, and then removing the `const` if they are indeed "variable".  

Also, would `all_empty_rows` be a more descriptive name? `empty_input` implies `input.size() == 0`, which isn't the case.

I think you've made the right call on using `count_if()`; I think the intention is captured better here. I'm sure `reduce()` would have worked too, but not as clear in intention. 
`thrust::all_of()` would have fit best, but [it is slow](https://github.com/NVIDIA/thrust/issues/1016).

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 3]
```diff
@@ -351,13 +351,17 @@ std::unique_ptr<column> percentile_approx(tdigest_column_view const& input,
...
+                                      input.size_begin(),
+                                      input.size_begin() + input.size(),
+                                      [] __device__(int x) { return x == 0; }) == input.size();
```
Comment (mythrocks) [inline cpp/src/quantiles/tdigest/tdigest.cu:357]:
```suggestion
                                      [] __device__(auto const x) { return x == 0; }) == input.size();
```
We could have used `size_type` as well, but `auto` would guarantee a match.

File: cpp/tests/quantiles/percentile_approx_test.cu [hunk 4]
```diff
@@ -405,8 +405,7 @@ TEST_F(PercentileApproxTest, EmptyInput)
                             3,
                             cudf::test::detail::make_null_mask(nulls.begin(), nulls.end()));
 
-  // TODO: change percentile_approx to produce sanitary list outputs for this case.
-  CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(*result, *expected);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*result, *expected);
```
Comment (mythrocks) [inline cpp/tests/quantiles/percentile_approx_test.cu:408]: :clap:

**Review by mythrocks — CHANGES_REQUESTED** [review]
Looks pretty good! Some minor nitpicks.

File: cpp/src/quantiles/tdigest/tdigest.cu (hunk 2 above)
Comment (mythrocks) [reply cpp/src/quantiles/tdigest/tdigest.cu:354]: > I think you've made the right call on using `count_if()`...

Ah, I see that @ttnghia was here already. :p

**Review by ttnghia — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
Ship it!

**Review by nvdbaranec — APPROVED** [review]

**Comment by nvdbaranec** [comment]
@gpucibot merge

**Pull request merged**